src/python_routing_v02/fast_reach/mc_reach.pyx

```diff
@@ -172,18 +175,49 @@ cpdef object compute_network(int nsteps, list reaches, dict connections,
     # rows: indexed by data_idx
     cdef float[:,::1] flowveldepth = np.zeros((data_idx.shape[0], nsteps * 3), dtype='float32')
 
+    # Pseudocode: LOOP ON Upstream Inflowers
+        # to pre-fill FlowVelDepth
+        # fill_index = list_of_all_segments_sorted -- .i.e, data_idx -- .index(upstream_tw_id)
+        # # FlowVelDepth[fill_index]['flow'] = UpstreamOutflows[upstream_tw_id]['flow']
+        # # FlowVelDepth[fill_index]['depth'] = UpstreamOutflows[upstream_tw_id]['depth']
+
+    # for ts in flowveldepth:
+        # print(f"{list(ts)}")
+
+    cdef set fill_index_mask = set()
+    # fill_index_mask is filled in the explicit loop below, which is
+    # identical to the following comprehension. But we use the explicit loop, because it
+    # is more transparent (and therefore optimizable) to cython.
+    # cdef set fill_index_mask = set([upstream_results[upstream_tw_id]["position_index"] for upstream_tw_id in upstream_results])
+    #print(f"{fill_index_mask}")
+
+    for upstream_tw_id in upstream_results:
+        fill_index = upstream_results[upstream_tw_id]["position_index"]
+        fill_index_mask.add(upstream_results[upstream_tw_id]["position_index"])
+        # print(f"{upstream_results[upstream_tw_id]['results']}")
+        # print(f"filling the {fill_index} row:")
+        # print(f"{list(flowveldepth[fill_index])}")
+        for idx, val in enumerate(upstream_results[upstream_tw_id]["results"]):
+            flowveldepth[fill_index][idx] = val
+        # TODO: Identify a more efficient ways potentially to handle this array filling
+        # The following may be options:
+        # flowveldepth[fill_index] = upstream_results[upstream_tw_id]["results"]
+        # flowveldepth[fill_index, :] = upstream_results[upstream_tw_id]["results"]
+        # print(f"Now filled, it contains:")
+        # print(f"{list(flowveldepth[fill_index])}")
+
```

Comment:
These lines read the values from the array provided in the call to populate the flowveldepth array with upstream inflows. this permits a very simple pattern for executing the computation in parallel, cascading subnetworks. An alternative approach would be to create a separate array for the upstream inflows, but this would require additional logic in the nogil section to select the correct source for upstream values depending on whether the `qup` and `quc` values were internal or external to the current subnetwork.